File: heapSort.py
def swap(a, i, j):
    t = a[i]
    a[i] = a[j]
    a[j] = t

def maxHeapify(a, i, heapSize):
    left = 2 * i + 1
    right = 2 * i + 2
    if left < heapSize and a[i]<a[left]:
        maxloc = left
    else:
        maxloc = i

    if right < heapSize and a[maxloc]<a[right]:
        maxloc = right

    if maxloc != i:
        swap(a, i, maxloc)
        maxHeapify(a, maxloc, heapSize)

def buildHeap(a):
    n = len(a)
    for i in range(n/2, -1, -1):
        maxHeapify(a, i, n)

def heapSort(a):
    buildHeap(a)
    n = len(a)
    heapSize = n
    #print (a)
    for i in range(n-1, 0, -1):
        swap(a, 0, i)
        heapSize = heapSize-1
        maxHeapify(a, 0, heapSize)

A = [6, 5, 3, 1, 8, 7, 2, 4]
print (A)
heapSort(A)
print (A)

File: sort_cpp/heapSort.cpp
#include <iostream>
using namespace std;

void swap(int *A, int i, int j){
  int t = A[i];
  A[i] = A[j];
  A[j] = t;
};

void maxHeapify(int *A, int i, int heapSize){
  int left = 2*i+1; int right = 2*i+2;
  int maxloc;
  if ((left<heapSize) and (A[i]<A[left])){
    maxloc = left;
  }
  else{
    maxloc = i;
  }
  if ((right<heapSize) and (A[maxloc]<A[right])){
    maxloc = right;
  }
  if (maxloc!=i){
    swap(A, i, maxloc);
    maxHeapify(A, maxloc, heapSize);
  }
};

void buildMaxHeap(int *A, int n){
  for (int i=n/2; i>=0; i--){
    maxHeapify(A, i, n);
  }
};

void heapSort(int *A, int n){
  buildMaxHeap(A, n);
  int heapSize = n;
  for (int i=n-1;i>=0;i--){
    swap(A, 0, i);
    heapSize --;
    maxHeapify(A, 0, heapSize);
  }
};

int main()
{
  int A[10] = {1,2,9,4,5,10,7,6,8,3};
  int n = 10;
  for (int i=0; i<n; i++){
    cout << A[i] << endl;
  }

  heapSort(A, n);

  for (int i=0; i<n; i++){
    cout << A[i] << endl;
  }
}
